File: client/lib/features/products/data layer/models/product_model.dart
import '../../../authorisation/data layer/models/user_model.dart';
import '../../../authorisation/domain layer/entities/user.dart';
import '../../domain layer/entities/product.dart';

class ProductModel extends Product {
  ProductModel(
    String id,
    List<String> productPictures,
    String name,
    String? description,
    num? priceBeforeReduction,
    num priceAfterReduction,
    num quantity,
    DateTime expirationDate,
    List<DateTime>? recoveryDate,
    User productOwner,
    Location location,
    bool expired,
    DateTime createdAt,
  ) : super(
          id,
          productPictures,
          name,
          description,
          priceBeforeReduction,
          priceAfterReduction,
          quantity,
          expirationDate,
          recoveryDate,
          productOwner,
          location,
          expired,
          createdAt,
        );

  factory ProductModel.fromJson(Map<String, dynamic> json) {
    return ProductModel(
      json['_id'] ?? '',
      List<String>.from(json['productPictures'] ?? []),
      json['name'] ?? '',
      json['description'],
      json['priceBeforeReduction'],
      json['priceAfterReduction'] ?? 0,
      json['quantity'] ?? 0,
      DateTime.parse(json['expirationDate'] ?? ''),
      List<DateTime>.from(
          json['recoveryDate'].map((date) => DateTime.parse(date))),
      UserModel.fromJson(json['productOwner'] as Map<String, dynamic>),
      Location(
        json['location'] != null
            ? (json['location']['coordinates'] as List<dynamic>)
                .map((e) => num.parse(e.toString()))
                .toList()
            : [0.0, 0.0],
      ),
      json['expired'] ?? false,
      DateTime.parse(json['createdAt'] ?? ''),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'productPictures': productPictures,
      'name': name,
      'description': description,
      'priceBeforeReduction': priceBeforeReduction,
      'priceAfterReduction': priceAfterReduction,
      'quantity': quantity,
      'expirationDate': expirationDate.toIso8601String(),
      'recoveryDate':
          recoveryDate?.map((date) => date?.toIso8601String()).toList(),
      'productOwner': (productOwner as UserModel).toJson(),
      'location': {
        'coordinates': location.coordinates,
      },
      'expired': expired,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

File: server/models/productModel.js
const mongoose = require("mongoose");
const validator = require("validator");

const productSchema = mongoose.Schema({
  productPictures: {
    type: [String],
    default: `defaultProduct${Math.floor(Math.random() * 4) + 1}.jpg`,
  },
  name: {
    type: String,
    required: [true, "Veuillez fournir un nom pour le produit"],
  },
  description: {
    type: String,
  },
  priceBeforeReduction: {
    type: Number,
  },
  priceAfterReduction: {
    type: Number,
    required: [true, "Veuillez fournir un prix pour le produit"],
  },
  quantity: {
    type: Number,
    required: [true, "Veuillez fournir une quantité pour le produit"],
  },
  expirationDate: {
    type: Date,
    required: [true, "Veuillez fournir une date d'expiration pour le produit"],
    default: () => {
      const endOfDay = new Date();
      endOfDay.setHours(23, 59, 59, 999);
      return endOfDay;
    },
  },
  recoveryDate: {
    type: [Date],
    default: () => {
      const sixPM = new Date();
      sixPM.setHours(18, 0, 0, 0);
      const eightPM = new Date();
      eightPM.setHours(20, 0, 0, 0);
      return [sixPM, eightPM];
    },
  },
  productOwner: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "User",
  },
  location: {
    type: {
      type: String,
      enum: ["Point"], // 'location.type' must be 'Point'
      required: true,
    },
    coordinates: {
      type: [Number],
      required: true,
    },
  },
  Expired: {
    type: Boolean,
    default: false,
  },
  createdAt: {
    type: Date,
    default: Date.now(),
  },
});
productSchema.pre(/^find/, function (next) {
  this.populate(
    "productOwner",
    "-password -passwordResetCode -passwordResetExpires -accountStatus -activeAccountToken -activeAccountTokenExpires"
  );
  next();
});
productSchema.index({ location: "2dsphere" });

const Product = mongoose.model("Product", productSchema);

module.exports = Product;
